app/src/main/java/com/example/myapplication/vm/LoginViewModel.kt

```diff
@@ -0,0 +1,46 @@
+package com.example.myapplication.vm
+
+import android.app.Application
+import android.content.Context
+import android.content.SharedPreferences
+import androidx.lifecycle.AndroidViewModel
+import androidx.lifecycle.MutableLiveData
+
+class LoginViewModel (app: Application): AndroidViewModel(app) {
+
+    val email = MutableLiveData<String?>()
+    val password = MutableLiveData<String?>()
+    val toast = MutableLiveData<Boolean>()
+    val save = MutableLiveData<Boolean>()
+    val error = MutableLiveData<Boolean>()
+    val sharedPreferences: SharedPreferences = app.applicationContext.getSharedPreferences("PREFERENCE_NAME", Context.MODE_PRIVATE)
+
+
+    fun nullValidation() {
```

Comment:
Also here, you aren't validating that the user is null, but instead that the user is valid! Think again in what this method does, and what name would be more proper.